File: CinePlayer/CinePlayer/Services/SubtitleService.swift
import Foundation

// MARK: - SubtitleService
/// Searches and downloads subtitles from OpenSubtitles.com (free REST API).
/// API documentation: https://opensubtitles.stoplight.io/docs/opensubtitles-api
/// Registration: https://www.opensubtitles.com/en/users/sign_up (free account)
/// Users must register for a free API key at: https://www.opensubtitles.com/en/consumers
final class SubtitleService {
    static let shared = SubtitleService()

    private let baseURL = "https://api.opensubtitles.com/api/v1"
    private let session = URLSession.shared

    private var bearerToken: String?
    private var tokenExpiry: Date?

    private var apiKey: String { AppSettings.shared.openSubtitlesAPIKey }
    private var username: String { AppSettings.shared.openSubtitlesUsername }
    private var password: String { AppSettings.shared.openSubtitlesPassword }

    private init() {}

    // MARK: - Public API

    /// Auto-searches and downloads the best subtitle for a media item.
    /// Uses OpenSubtitles hash search first, then falls back to title search.
    func findAndDownloadSubtitle(for item: MediaItem) async throws -> SubtitleTrack? {
        guard !apiKey.isEmpty else { throw SubtitleError.noAPIKey }

        let language = AppSettings.shared.preferredSubtitleLanguage

        // Try hash-based search first (most accurate)
        if let result = try? await searchByHash(fileURL: item.fileURL, language: language) {
            return try await downloadSubtitle(result, for: item)
        }

        // Fall back to title + year search
        let query: String
        if item.mediaType == .episode,
           let seriesName = item.seriesName,
           let season = item.seasonNumber,
           let episode = item.episodeNumber {
            let results = try await searchByEpisode(
                query: seriesName,
                season: season,
                episode: episode,
                language: language
            )
            if let best = results.first {
                return try await downloadSubtitle(best, for: item)
            }
            return nil
        } else {
            query = item.title
            let year = item.releaseYear
            let results = try await searchByTitle(query: query, year: year, language: language)
            guard let best = results.first else { return nil }
            return try await downloadSubtitle(best, for: item)
        }
    }

    // MARK: - Hash Search

    /// Compute OpenSubtitles file hash and search for matching subtitles
    private func searchByHash(fileURL: URL, language: String) async throws -> SubtitleSearchResult? {
        let hash = try computeHash(fileURL: fileURL)
        let size = try fileSize(fileURL: fileURL)

        var components = URLComponents(string: baseURL + "/subtitles")!
        components.queryItems = [
            URLQueryItem(name: "moviehash", value: hash),
            URLQueryItem(name: "moviehash_match", value: "only"),
            URLQueryItem(name: "languages", value: language),
        ]

        let results = try await getAuthenticated(url: components.url!, as: SubtitleSearchResponse.self)
        _ = size // hash search implicitly uses file size
        return results.data.first
    }

    // MARK: - Title Search

    func searchByTitle(query: String, year: Int? = nil, language: String) async throws -> [SubtitleSearchResult] {
        var components = URLComponents(string: baseURL + "/subtitles")!
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "query", value: query),
            URLQueryItem(name: "type", value: "movie"),
            URLQueryItem(name: "languages", value: language),
        ]
        if let y = year {
            queryItems.append(URLQueryItem(name: "year", value: String(y)))
        }
        components.queryItems = queryItems
        let results = try await getAuthenticated(url: components.url!, as: SubtitleSearchResponse.self)
        return results.data
    }

    func searchByEpisode(
        query: String,
        season: Int,
        episode: Int,
        language: String
    ) async throws -> [SubtitleSearchResult] {
        var components = URLComponents(string: baseURL + "/subtitles")!
        components.queryItems = [
            URLQueryItem(name: "query", value: query),
            URLQueryItem(name: "type", value: "episode"),
            URLQueryItem(name: "season_number", value: String(season)),
            URLQueryItem(name: "episode_number", value: String(episode)),
            URLQueryItem(name: "languages", value: language),
        ]
        let results = try await getAuthenticated(url: components.url!, as: SubtitleSearchResponse.self)
        return results.data
    }

    // MARK: - Download

    private func downloadSubtitle(_ result: SubtitleSearchResult, for item: MediaItem) async throws -> SubtitleTrack {
        // Request download link
        let requestBody = ["file_id": result.attributes.files.first?.fileID ?? 0]
        let downloadInfo = try await postAuthenticated(
            endpoint: "/download",
            body: requestBody,
            as: SubtitleDownloadResponse.self
        )

        // Download the subtitle file
        let (data, _) = try await session.data(from: URL(string: downloadInfo.link)!)

        // Save to disk alongside the video file (or in subtitles cache)
        let subtitleDir = item.fileURL.deletingLastPathComponent()
            .appendingPathComponent("Subtitles", isDirectory: true)
        try FileManager.default.createDirectory(at: subtitleDir, withIntermediateDirectories: true)

        let lang = result.attributes.language ?? AppSettings.shared.preferredSubtitleLanguage
        let ext = downloadInfo.fileName.components(separatedBy: ".").last ?? "srt"
        let baseName = item.fileURL.deletingPathExtension().lastPathComponent
        let saveURL = subtitleDir.appendingPathComponent("\(baseName).\(lang).\(ext)")
        try data.write(to: saveURL)

        return SubtitleTrack(
            language: languageName(from: lang),
            languageCode: lang,
            fileURL: saveURL,
            isEmbedded: false,
            source: .downloaded
        )
    }

    // MARK: - Authentication

    private func ensureAuthenticated() async throws {
        // Check if we have a valid token
        if let token = bearerToken,
           let expiry = tokenExpiry, expiry > Date() {
            return
        }
        // If no username/password, use API key only (guest mode for OpenSubtitles v3)
        guard !username.isEmpty && !password.isEmpty else {
            // API key only (limited downloads per day, no login needed)
            return
        }
        try await login()
    }

    private func login() async throws {
        let body: [String: Any] = ["username": username, "password": password]
        let response = try await post(
            endpoint: "/login",
            body: body,
            headers: ["Api-Key": apiKey],
            as: LoginResponse.self
        )
        bearerToken = response.token
        tokenExpiry = Date().addingTimeInterval(24 * 60 * 60) // 24 hours
    }

    func logout() async throws {
        guard bearerToken != nil else { return }
        _ = try? await deleteAuthenticated(endpoint: "/logout")
        bearerToken = nil
        tokenExpiry = nil
    }

    // MARK: - OpenSubtitles Hash Algorithm
    // Reference: https://trac.opensubtitles.org/projects/opensubtitles/wiki/HashSourceCodes

    private func computeHash(fileURL: URL) throws -> String {
        let file = try FileHandle(forReadingFrom: fileURL)
        defer { file.closeFile() }

        let fileSize = try self.fileSize(fileURL: fileURL)
        guard fileSize >= 131072 else { throw SubtitleError.fileTooSmall }

        var hash: UInt64 = UInt64(fileSize)

        // Read first 64KB
        let chunkSize = 65536
        let firstChunk = file.readData(ofLength: chunkSize)
        hash = xorHash(data: firstChunk, initial: hash)

        // Read last 64KB
        file.seek(toFileOffset: UInt64(fileSize) - UInt64(chunkSize))
        let lastChunk = file.readData(ofLength: chunkSize)
        hash = xorHash(data: lastChunk, initial: hash)

        return String(format: "%016llx", hash)
    }

    private func xorHash(data: Data, initial: UInt64) -> UInt64 {
        var hash = initial
        let count = data.count / 8
        data.withUnsafeBytes { ptr in
            let words = ptr.bindMemory(to: UInt64.self)
            for i in 0..<count {
                hash = hash &+ words[i].littleEndian
            }
        }
        return hash
    }

    private func fileSize(fileURL: URL) throws -> Int64 {
        let attrs = try FileManager.default.attributesOfItem(atPath: fileURL.path)
        return (attrs[.size] as? Int64) ?? 0
    }

    // MARK: - Network Helpers

    private func getAuthenticated<T: Decodable>(url: URL, as type: T.Type) async throws -> T {
        try await ensureAuthenticated()
        var request = URLRequest(url: url)
        request.setValue(apiKey, forHTTPHeaderField: "Api-Key")
        request.setValue("CinePlayer v1.0", forHTTPHeaderField: "User-Agent")
        if let token = bearerToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func postAuthenticated<T: Decodable>(
        endpoint: String,
        body: [String: Any],
        as type: T.Type
    ) async throws -> T {
        try await ensureAuthenticated()
        return try await post(
            endpoint: endpoint,
            body: body,
            headers: buildAuthHeaders(),
            as: T.self
        )
    }

    private func deleteAuthenticated(endpoint: String) async throws {
        try await ensureAuthenticated()
        var request = URLRequest(url: URL(string: baseURL + endpoint)!)
        request.httpMethod = "DELETE"
        for (key, value) in buildAuthHeaders() {
            request.setValue(value, forHTTPHeaderField: key)
        }
        _ = try await session.data(for: request)
    }

    private func post<T: Decodable>(
        endpoint: String,
        body: [String: Any],
        headers: [String: String],
        as type: T.Type
    ) async throws -> T {
        var request = URLRequest(url: URL(string: baseURL + endpoint)!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func buildAuthHeaders() -> [String: String] {
        var headers: [String: String] = [
            "Api-Key": apiKey,
            "User-Agent": "CinePlayer v1.0",
        ]
        if let token = bearerToken {
            headers["Authorization"] = "Bearer \(token)"
        }
        return headers
    }

    private func validateResponse(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { return }
        switch http.statusCode {
        case 200...299: return
        case 401: throw SubtitleError.unauthorized
        case 429: throw SubtitleError.rateLimited
        default: throw SubtitleError.serverError(http.statusCode)
        }
    }

    private func languageName(from code: String) -> String {
        let locale = Locale(identifier: "en")
        return locale.localizedString(forLanguageCode: code) ?? code
    }

    // MARK: - Error

    enum SubtitleError: LocalizedError {
        case noAPIKey
        case fileTooSmall
        case unauthorized
        case rateLimited
        case serverError(Int)

        var errorDescription: String? {
            switch self {
            case .noAPIKey: return "OpenSubtitles API key not configured. Add it in Settings."
            case .fileTooSmall: return "File too small for hash-based search."
            case .unauthorized: return "Invalid OpenSubtitles credentials."
            case .rateLimited: return "OpenSubtitles rate limit reached. Try again later."
            case .serverError(let code): return "Server error: \(code)"
            }
        }
    }
}

// MARK: - OpenSubtitles API Models

struct SubtitleSearchResponse: Codable {
    let data: [SubtitleSearchResult]
}

struct SubtitleSearchResult: Codable, Identifiable {
    let id: String
    let attributes: SubtitleAttributes
}

struct SubtitleAttributes: Codable {
    let language: String?
    let downloadCount: Int?
    let newDownloadCount: Int?
    let hearingImpaired: Bool?
    let hd: Bool?
    let fps: Double?
    let votes: Int?
    let ratings: Double?
    let fromTrusted: Bool?
    let foreignParts: Bool?
    let uploadDate: String?
    let files: [SubtitleFile]

    enum CodingKeys: String, CodingKey {
        case language
        case downloadCount = "download_count"
        case newDownloadCount = "new_download_count"
        case hearingImpaired = "hearing_impaired"
        case hd, fps, votes, ratings
        case fromTrusted = "from_trusted"
        case foreignParts = "foreign_parts"
        case uploadDate = "upload_date"
        case files = "files"
    }
}

struct SubtitleFile: Codable {
    let fileID: Int
    let fileName: String?

    enum CodingKeys: String, CodingKey {
        case fileID = "file_id"
        case fileName = "file_name"
    }
}

struct SubtitleDownloadResponse: Codable {
    let link: String
    let fileName: String
    let requests: Int
    let allowed: Int
    let remaining: Int
    let message: String?
    let resetTime: String?

    enum CodingKeys: String, CodingKey {
        case link, requests, allowed, remaining, message
        case fileName = "file_name"
        case resetTime = "reset_time"
    }
}

struct LoginResponse: Codable {
    let token: String
    let status: Int?
    let user: LoginUser?

    struct LoginUser: Codable {
        let allowedDownloads: Int
        let level: String
        let userId: Int
        let extInstalled: Bool
        let vip: Bool

        enum CodingKeys: String, CodingKey {
            case allowedDownloads = "allowed_downloads"
            case level
            case userId = "user_id"
            case extInstalled = "ext_installed"
            case vip
        }
    }
}

File: CinePlayerAndroid/app/src/main/java/com/cineplayer/android/services/SubtitleService.kt
package com.cineplayer.android.services

import android.content.Context
import android.net.Uri
import com.cineplayer.android.data.AppSettings
import com.cineplayer.android.models.MediaItem
import com.cineplayer.android.models.MediaType
import com.cineplayer.android.models.SubtitleSource
import com.cineplayer.android.models.SubtitleTrack
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.io.File
import java.io.FileOutputStream
import java.io.RandomAccessFile
import java.net.URL

data class OSLoginRequest(@SerializedName("username") val username: String, @SerializedName("password") val password: String)
data class OSLoginResponse(@SerializedName("token") val token: String? = null)
data class OSSearchResponse(@SerializedName("data") val data: List<OSSubtitleResult> = emptyList())
data class OSSubtitleResult(
    @SerializedName("id") val id: String = "",
    @SerializedName("attributes") val attributes: OSSubtitleAttributes = OSSubtitleAttributes()
)
data class OSSubtitleAttributes(
    @SerializedName("language") val language: String = "",
    @SerializedName("download_count") val downloadCount: Int = 0,
    @SerializedName("moviehash_match") val moviehashMatch: Boolean = false,
    @SerializedName("files") val files: List<OSSubtitleFile> = emptyList()
)
data class OSSubtitleFile(@SerializedName("file_id") val fileId: Int = 0, @SerializedName("file_name") val fileName: String = "")
data class OSDownloadRequest(@SerializedName("file_id") val fileId: Int)
data class OSDownloadResponse(@SerializedName("link") val link: String = "", @SerializedName("file_name") val fileName: String = "", @SerializedName("remaining") val remaining: Int = 0)

interface OpenSubtitlesApi {
    @POST("login")
    suspend fun login(@Body request: OSLoginRequest, @Header("Api-Key") apiKey: String): OSLoginResponse
    @GET("subtitles")
    suspend fun searchByHash(@Query("moviehash") hash: String, @Query("languages") language: String, @Header("Api-Key") apiKey: String, @Header("Authorization") token: String = ""): OSSearchResponse
    @GET("subtitles")
    suspend fun searchByTitle(@Query("query") query: String, @Query("languages") language: String, @Query("year") year: Int? = null, @Header("Api-Key") apiKey: String, @Header("Authorization") token: String = ""): OSSearchResponse
    @GET("subtitles")
    suspend fun searchByEpisode(@Query("query") query: String, @Query("season_number") season: Int, @Query("episode_number") episode: Int, @Query("languages") language: String, @Header("Api-Key") apiKey: String, @Header("Authorization") token: String = ""): OSSearchResponse
    @POST("download")
    suspend fun download(@Body request: OSDownloadRequest, @Header("Api-Key") apiKey: String, @Header("Authorization") token: String = ""): OSDownloadResponse
}

class SubtitleService private constructor(private val context: Context) {
    /** Thrown when the OpenSubtitles daily download quota (HTTP 429) is exceeded. */
    class RateLimitException : Exception("OpenSubtitles daily download limit reached")

    private val settings = AppSettings.getInstance(context)
    private var authToken: String = ""

    private val api: OpenSubtitlesApi = Retrofit.Builder()
        .baseUrl("https://api.opensubtitles.com/api/v1/")
        .client(OkHttpClient.Builder()
            .addInterceptor(HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BASIC })
            .build())
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(OpenSubtitlesApi::class.java)

    companion object {
        @Volatile private var instance: SubtitleService? = null
        fun getInstance(context: Context): SubtitleService =
            instance ?: synchronized(this) { instance ?: SubtitleService(context.applicationContext).also { instance = it } }
    }

    suspend fun login(): Boolean {
        val apiKey = settings.openSubtitlesApiKey.ifBlank { return false }
        val username = settings.openSubtitlesUsername
        val password = settings.openSubtitlesPassword
        if (username.isBlank() || password.isBlank()) return false
        return try {
            val response = api.login(OSLoginRequest(username, password), apiKey)
            authToken = response.token?.let { "Bearer $it" } ?: ""
            authToken.isNotBlank()
        } catch (_: Exception) { false }
    }

    suspend fun findAndDownloadSubtitle(item: MediaItem): SubtitleTrack? = withContext(Dispatchers.IO) {
        val apiKey = settings.openSubtitlesApiKey.ifBlank { return@withContext null }
        val language = settings.preferredSubtitleLanguage

        findLocalSubtitle(item)?.let { return@withContext it }

        if (item.mediaType == MediaType.MOVIE) {
            val path = Uri.parse(item.fileUri).path
            if (path != null) {
                computeHash(File(path))?.let { hash ->
                    try {
                        val results = api.searchByHash(hash, language, apiKey, authToken)
                        val best = results.data.maxByOrNull { (if (it.attributes.moviehashMatch) 1000 else 0) + it.attributes.downloadCount }
                        if (best != null) downloadAndSave(best, item, apiKey)?.let { return@withContext it }
                    } catch (_: Exception) {}
                }
            }
        }

        if (item.mediaType == MediaType.EPISODE && item.seriesName != null && item.seasonNumber != null && item.episodeNumber != null) {
            try {
                val results = api.searchByEpisode(item.seriesName!!, item.seasonNumber!!, item.episodeNumber!!, language, apiKey, authToken)
                results.data.maxByOrNull { it.attributes.downloadCount }?.let { downloadAndSave(it, item, apiKey)?.let { t -> return@withContext t } }
            } catch (_: Exception) {}
        }

        try {
            val results = api.searchByTitle(item.title, language, item.releaseYear, apiKey, authToken)
            results.data.maxByOrNull { it.attributes.downloadCount }?.let { downloadAndSave(it, item, apiKey) }
        } catch (_: Exception) { null }
    }

    private fun findLocalSubtitle(item: MediaItem): SubtitleTrack? {
        val filePath = Uri.parse(item.fileUri).path ?: return null
        val dir = File(filePath).parentFile ?: return null
        val baseName = File(filePath).nameWithoutExtension
        val exts = listOf("srt", "vtt", "ass", "ssa")
        val candidate = dir.listFiles { f ->
            val name = f.nameWithoutExtension
            f.extension.lowercase() in exts && (name == baseName || name.startsWith("$baseName.") || name.startsWith("${baseName}_"))
        }?.firstOrNull() ?: return null
        val langCode = candidate.nameWithoutExtension.removePrefix(baseName).trimStart('.', '_').lowercase().take(3).ifBlank { "en" }
        return SubtitleTrack(language = langCode.uppercase(), languageCode = langCode, filePath = candidate.absolutePath, source = SubtitleSource.LOCAL)
    }

    private suspend fun downloadAndSave(result: OSSubtitleResult, item: MediaItem, apiKey: String): SubtitleTrack? {
        val fileId = result.attributes.files.firstOrNull()?.fileId ?: return null
        return try {
            val resp = api.download(OSDownloadRequest(fileId), apiKey, authToken)
            val link = resp.link.ifBlank { return null }
            val videoPath = Uri.parse(item.fileUri).path ?: return null
            val subtitlesDir = File(File(videoPath).parentFile, "Subtitles").also { it.mkdirs() }
            val fileName = resp.fileName.ifBlank { "${item.title}.${result.attributes.language}.srt" }
            val dest = File(subtitlesDir, fileName)
            FileOutputStream(dest).use { it.write(URL(link).readBytes()) }
            SubtitleTrack(language = result.attributes.language.uppercase(), languageCode = result.attributes.language.lowercase(), filePath = dest.absolutePath, source = SubtitleSource.DOWNLOADED)
        } catch (e: retrofit2.HttpException) {
            if (e.code() == 429) throw RateLimitException()
            null
        } catch (_: Exception) { null }
    }

    private fun computeHash(file: File): String? {
        if (!file.exists()) return null
        return try {
            val fileSize = file.length()
            val chunkSize = minOf(65536L, fileSize).toInt()
            var hash = fileSize
            RandomAccessFile(file, "r").use { raf ->
                val buf = ByteArray(chunkSize)
                raf.seek(0); raf.read(buf)
                for (i in buf.indices step 8) { hash = (hash + readLong(buf, i)) and Long.MAX_VALUE }
                raf.seek(maxOf(0, fileSize - chunkSize)); raf.read(buf)
                for (i in buf.indices step 8) { hash = (hash + readLong(buf, i)) and Long.MAX_VALUE }
            }
            "%016x".format(hash)
        } catch (_: Exception) { null }
    }

    private fun readLong(bytes: ByteArray, offset: Int): Long {
        var v = 0L
        for (i in 0 until minOf(8, bytes.size - offset)) v = v or ((bytes[offset + i].toLong() and 0xFF) shl (i * 8))
        return v
    }
}
